data-prepper-plugins/otel-proto-common/src/main/java/org/opensearch/dataprepper/plugins/otel/codec/OTelMetricDecoder.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.otel.codec;
+
+import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceRequest;
+import org.opensearch.dataprepper.model.codec.ByteDecoder;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.event.JacksonEvent;
+import org.opensearch.dataprepper.model.metric.Metric;
+
+import org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec;
+
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.function.Consumer;
+
+
+public class OTelMetricDecoder implements ByteDecoder {
+    private final OTelProtoCodec.OTelProtoDecoder otelProtoDecoder;
+    public OTelMetricDecoder() {
+        otelProtoDecoder = new OTelProtoCodec.OTelProtoDecoder();
+    }
+    public void parse(InputStream inputStream, Consumer<Record<Event>> eventConsumer) throws IOException {
+        ExportMetricsServiceRequest request = ExportMetricsServiceRequest.parseFrom(inputStream);
+        AtomicInteger droppedCounter = new AtomicInteger(0);
+        Collection<Record<? extends Metric>> records =
+            otelProtoDecoder.parseExportMetricsServiceRequest(request, droppedCounter, OTelProtoCodec.DEFAULT_EXPONENTIAL_HISTOGRAM_MAX_ALLOWED_SCALE, true, true, false);
```

Comment:
If I understand this correctly, the pipeline will have different behavior depending on the buffer type. With the in-memory buffer, we place the whole request in the buffer and it is the first object out of the pipeline. With an external buffer, it appears that we are now providing the actual metrics as the first events in the buffer.

I think the pipelines should be consistent between the two buffers. That is, you can swap your buffer without having to change your pipeline structure.

Reply:
I can't think of any way to make it 100% same. But the YAML file can stay same and it works same.